stepper-lib/src/air/call/utils.rs

```diff
@@ -40,9 +40,34 @@ pub(super) fn set_local_call_result<'i>(
     match output {
         CallOutput::Scalar(name) => {
             match exec_ctx.data_cache.entry(name.to_string()) {
-                Vacant(entry) => entry.insert(AValue::JValueRef(result)),
-                Occupied(entry) => return Err(MultipleVariablesFound(entry.key().clone())),
+                Vacant(entry) => {
+                    entry.insert(AValue::JValueRef(result));
+                }
+                Occupied(mut entry) => {
+                    // check that current execution flow is inside a fold block
+                    if exec_ctx.met_folds.is_empty() {
+                        // shadowing is allowed only inside fold blocks
+                        return Err(MultipleVariablesFound(entry.key().clone()));
+                    }
+
+                    match entry.get() {
+                        AValue::JValueRef(_) => {}
+                        // shadowing is allowed only for scalar values
+                        _ => return Err(MultipleVariablesFound(entry.key().clone())),
```

Comment:
I think this deserves a separate error with a special error message like `"ShadowError: Accumulator shadowing is not supported: {}", entry.key`